python/ray/data/_internal/execution/backpressure_policy.py

```diff
@@ -113,3 +134,70 @@ def can_run(self, op: "PhysicalOperator") -> bool:
+
+
+class StreamingOutputBackpressurePolicy(BackpressurePolicy):
+    """A backpressure policy that throttles the streaming outputs of the `DataOpTask`s.
+
+    The are 2 levels of configs to control the behavior:
+    - At the Ray Core level, we use
+      `MAX_BLOCKS_IN_GENERATOR_BUFFER` to limit the number of blocks buffered in
+      the streaming generator of each OpDataTask. When it's reached, the task will
+      be blocked at `yield`.
+    - At the Ray Data level, we use
+      `MAX_BLOCKS_IN_GENERATOR_BUFFER` to limit the number of blocks buffered in the
+      output queue of each operator. When it's reached, we'll stop reading from the
+      streaming generators of the op's tasks, and thus trigger backpressure at the
+      Ray Core level.
+    """
+
+    # The max number of blocks that can be buffered at the streaming generator
+    # of each `DataOpTask`.
+    MAX_BLOCKS_IN_GENERATOR_BUFFER = 10
+    MAX_BLOCKS_IN_GENERATOR_BUFFER_BUFFER_CONFIG_KEY = (
+        "backpressure_policies.streaming_output.max_blocks_in_generator_buffer"
+    )
+    # The max number of blocks that can be buffered at the operator output queue
+    # (`OpState.outqueue`).
+    MAX_BLOCKS_IN_OP_OUTPUT_QUEUE = 20
+    MAX_BLOCKS_IN_OP_OUTPUT_QUEUE_CONFIG_KEY = (
+        "backpressure_policies.streaming_output.max_blocks_in_op_output_queue"
+    )
+
+    def __init__(self, topology: "Topology"):
+        data_context = ray.data.DataContext.get_current()
+        self._max_num_blocks_in_streaming_gen_buffer = data_context.get_config(
+            self.MAX_BLOCKS_IN_GENERATOR_BUFFER_BUFFER_CONFIG_KEY,
+            self.MAX_BLOCKS_IN_GENERATOR_BUFFER,
+        )
+        # The `_generator_backpressure_num_objects` parameter should be
+        # `2 * self._max_num_blocks_in_streaming_gen_buffer` because we yield
+        # 2 objects for each block: the block and the block metadata.
+        data_context._task_pool_data_task_remote_args[
+            "_generator_backpressure_num_objects"
+        ] = (2 * self._max_num_blocks_in_streaming_gen_buffer)
+
+        self._max_num_blocks_in_op_output_queue = data_context.get_config(
+            self.MAX_BLOCKS_IN_OP_OUTPUT_QUEUE_CONFIG_KEY,
+            self.MAX_BLOCKS_IN_OP_OUTPUT_QUEUE,
+        )
+
+    def calcuate_max_blocks_to_read_per_op(
+        self, topology: "Topology"
+    ) -> Dict["OpState", int]:
+        max_blocks_to_read_per_op: Dict["OpState", int] = {}
+        downstream_num_active_tasks = 0
+        for op, state in list(topology.items())[::-1]:
+            max_blocks_to_read_per_op[state] = (
+                self._max_num_blocks_in_op_output_queue - state.outqueue_num_blocks()
+            )
+            if downstream_num_active_tasks == 0:
+                # If all downstream operators are idle, it could be because no resources
+                # are available. In this case, we'll make sure to read at least one
+                # block to avoid deadlock.
```

Comment:
I am still unsure about this condition.

"it could be" -> meaning that there cases where the downstream operators are idle but it is not a deadlock? Can we guarantee that it is a deadlock?

Reply:
Actually isn't exponential rampup one of the cases where it's not actually a deadlock?

I guess it is okay to merge for now, but I think we should revisit this condition after 2.8.

Reply:
yes, this is not 100% percent accurate. so the current strategy is conservative, it only unblocks one block at a time. I'll add a TODO here.

Reply:
After a second thought, it seems that the only false positive case is when the executor just starts, the first op hasn't produced blocks for the second op to consume. 
For the exponential ramp-up case, if an op is backpressured due to the concurrency cap. `num_active_tasks` will be greater than 0.